src/microbe_stage/BackgroundPlane.cs

```diff
@@ -0,0 +1,249 @@
+    /// </summary>
+    public void SetBackground(Background background)
+    {
+        // TODO: skip duplicate background changes
+
+        for (int i = 0; i < 4; ++i)
+        {
+            // TODO: switch this loop away to reuse StringName instances if this causes significant allocations
+            currentBackgroundMaterial.SetShaderParameter($"layer{i:n0}", GD.Load<Texture2D>(background.Textures[i]));
+        }
+
+        backgroundParticles?.DetachAndQueueFree();
+
+        backgroundParticles = background.ParticleEffectScene.Instantiate<GpuParticles3D>();
+        backgroundParticles.Rotation = Rotation;
+        backgroundParticles.LocalCoords = false;
+
+        AddChild(backgroundParticles);
+
+        OnDisplayBackgroundParticlesChanged(Settings.Instance.DisplayBackgroundParticles);
+    }
+
+    public void SetWorldPosition(Vector2 position)
+    {
+        currentBackgroundMaterial.SetShaderParameter(worldPositionParameter, position);
+    }
+
+    public void SetVisibility(bool visible)
+    {
+        Visible = visible;
+
+        if (backgroundParticles != null)
+            OnDisplayBackgroundParticlesChanged(Settings.Instance.DisplayBackgroundParticles);
+    }
+
+    public void UpdateLightLevel(float lightLevel)
+    {
+        currentBackgroundMaterial.SetShaderParameter(lightLevelParameter, lightLevel);
+    }
+
+    protected override void Dispose(bool disposing)
+    {
+        if (disposing)
+        {
+            textureAlbedoParameter.Dispose();
+            blurAmountParameter.Dispose();
+            lightLevelParameter.Dispose();
+            distortionStrengthParameter.Dispose();
+            worldPositionParameter.Dispose();
+        }
+
+        base.Dispose(disposing);
+    }
+
+    private void OnDisplayBackgroundParticlesChanged(bool displayed)
+    {
+        if (backgroundParticles == null)
+        {
+            GD.PrintErr("BackgroundPlane didn't find background particles on settings change");
+            return;
+        }
+
+        backgroundParticles.Emitting = displayed;
+
+        if (displayed)
+        {
+            backgroundParticles.Show();
+        }
+        else
+        {
+            backgroundParticles.Hide();
+        }
+    }
+
+    private void OnBackgroundDistortionChanged(float value)
+    {
+        ApplyDistortionEffect();
+    }
+
+    private void ApplyDistortionEffect()
+    {
+        currentBackgroundMaterial.SetShaderParameter(distortionStrengthParameter,
+            Settings.Instance.MicrobeDistortionStrength.Value);
+    }
+
+    private void OnBackgroundBlurStrengthChanged(float value)
+    {
+        ApplyBlurEffect();
+    }
+
+    private void OnBackgroundBlurToggleChanged(bool value)
+    {
+        ApplyBlurEffect();
+    }
+
+    private void ApplyBlurEffect()
+    {
+        float blurStrength = Settings.Instance.MicrobeBackgroundBlurStrength;
+        bool enabled = blurStrength > 0;
+
+        if (blurEnabledLastTime == enabled)
```

Comment:
Doesn't this work incorrectly if blur strength is changed from 1.0 to 1.1 (i.e. it would early return and not apply the strength)?

Reply:
I meant to add that I suspect that due to the code in `OnBackgroundBlurStrengthChanged`